### PR TITLE
Decouple purge and hold

### PR DESCRIPTION
**Pull request by giuliano108**
* Purge and hold can be toggled on/off separately. Before this PR, purging had to be enabled for holding to work.
* `noop` is free from side-effects. `aptly_purge` should not write into `apt-mark` or `dpkg --set-selections` when running in noop mode.
* Packages are "un-held" when they no longer appear in the catalog or they stop being pinned to a specific version (i.e., they are set to "`ensure => present`").

This is a bit of a departure from the original `aptly_purge`. Should we change the name to something more descriptive, and maybe stop trying to have these changes merged upstream? I'm also thinking about how we are going to roll out this version (when/if it's ready). Different names might help us keeping both the old and new version in the Puppet tree...

### CODE REVIEW

File: spec/acceptance/02_holds_safely_spec.rb [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    ]
+    @package_versions = {}
+
+    hosts.each do |host|
+      install_package host, 'dict-jargon'
+      expect(check_for_package host, 'dictd').to be true
+      install_package host, 'fortunes'
+      expect(check_for_package host, 'fortunes-min').to be true
+      # same as `include package_purging::config`, saves a Puppet run
+      create_remote_file host, '/etc/apt/apt.conf.d/99always-purge', "APT::Get::Purge \"true\";\n";
```
Comment (oholiab) [inline spec/acceptance/02_holds_safely_spec.rb:47]: Just want to double check that we're sufficiently happy that this fundamental of the way this module works is in fact fundamental and we're not going to break all the tests in weird and wonderful ways later on...

I'm cool with it as long as we're fairly certain this won't suddenly be invalid.

File: spec/acceptance/02_holds_safely_spec.rb [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      # fortunes isn't
+      expect(packages_state['fortunes']).to eq('install')
+      # because the "purge" parameter defaults to false, the package is still installed
+      expect(package('fortunes')).to be_installed
+    end
+  end
+
+
+  RSpec.shared_examples 'aptly_purge (hold) noop' do |test_case|
+    it 'maeks no changes' do
```
Comment (oholiab) [inline spec/acceptance/02_holds_safely_spec.rb]: s/maeks/makes/

**Review by oholiab — APPROVED** [review]
lgtm, gonna think a bit on the test coverage just in case there are some edge cases but that shouldn't block you from shipping this :)

File: spec/acceptance/02_holds_safely_spec.rb (hunk 1 above)
Comment (giuliano108) [reply spec/acceptance/02_holds_safely_spec.rb:47]: The module provides an `package_purging::config` which is mentioned in the README.

`package_purging::config` is tested in `00_purges_safely_spec.rb`, everywhere else I'm hackishly creating the file by hand to save a couple of Puppet runs. Some context is available [here](https://github.com/hashbrowncipher/puppet-package_purging/pull/4#issuecomment-272715594).

**Review by solarkennedy — APPROVED** [review]

**Comment by hashbrowncipher** [comment]
Hey @giuliano108,

I'm looking to merge these commits upstream, and I'm having trouble understanding why you made the changes you did to noop mode.

The purpose of noop mode (in Puppet generally, and for this module specifically) is to give the reader useful insight into what would happen if the functionality were run "for real". To my reading (and do correct me if I'm wrong) running this PR in noop mode will produce no output at all. The reader gains no insight. It's as if the code isn't even present.

I'm guessing that you removed the noop functionality because aptly_purge writes into the apt-mark system even in noop mode. I view this behavior as unfortunate, but essential. Did you have another motivation that I haven't thought of? If so, I'd like to try to find a way to accommodate it.

Being able to gain insight in noop mode is extremely important for adoption of this module by new users, and for the safety of existing users. Users cannot adopt this module if they cannot test it, see what it would remove, and modify their Puppet catalog accordingly. And, with this PR, an existing user making changes to Package resources in their catalog has no way of, in a noop run, determining what other packages might be removed by aptly_purge.

If the user has some set of machines that they don't want to run package purging on, my recommended solution for them would be to enclose the aptly_purge resource itself in a conditional clause.

Previous iterations of package purging were bash or ruby scripts than ran outside of Puppet runs by analyzing json catalogs from /var/lib/puppet. I/we deemed them too hazardous to deploy in production, so that code languished. By the same token, I would deem it hazardous to run this PR in production. It's too easy for a user to (with aptly_purge's help) unwittingly remove package(s) they care about, and no affordances are made for them to test their change before it goes live.

Can we come up with a version of aptly_purge that contains your improvements and retains the functionality of noop mode? Would you be willing to merge such a change into Yelp's fork?

Josh

**Comment by giuliano108** [comment]
> I'm guessing that you removed the noop functionality because aptly_purge writes into the apt-mark system even in noop mode.

That is one of the reasons, but I also wanted to be able to disable/enable purging and holding individually, which I cannot do with a common `noop` toggle I think.

> And, with this PR, an existing user making changes to Package resources in their catalog has no way of, in a noop run, determining what other packages might be removed by aptly_purge.

You're right, I'm subverting the expected behaviour of `noop`. If I were to deploy automatic purging everywhere at work, would `noop` be a useful tool to gauge the safety of such a change? I'm not sure. I'd probably want to instrument the code and get the key data structures stored outside of the normal Puppet output, in a format that'd make it easy to run fleet-wide reports of "what would be purged/held".

> Can we come up with a version of aptly_purge that contains your improvements and retains the functionality of noop mode? Would you be willing to merge such a change into Yelp's fork?

Would separating `aptly_purge` from an hypothetical `aptly_hold` make sense? I'd like that, but I'm learning this stuff as I go and could be wrong. :)

Out of context, allow me to mention https://github.com/Yelp/puppet-package_purging/pull/3 _"package holding and aptitude issues"_. I sort of worked around the `provider => aptitude` problem, but might have hit a blocker with manifests that explicitly set `provider => apt` (I'm afraid `defaultfor :osfamily => :debian` doesn't work in the latter case).

Thanks for the feedback!

--
Giuliano

**Comment by hashbrowncipher** [comment]
You're quite welcome!

It sounds like it might make the most sense to split the type into two, so that noop retains its usefulness.
